### PR TITLE
Ultimate jobs fix

### PR DESCRIPTION
**Pull request by ssichynskyi**
this pull request fixes search_jobs() including issues:
#147
#144
#125 (at least one problem from comments there)
as well as adds distance feature

apart from that:
- fix some other tests
- fix formatting issues
- updates documentation and contributing guidelines

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,5 +1,5 @@
...
-LINKEDIN_PASSWORD="coolpasswordhere"
-TEST_PROFILE_ID="ACoAABQ11fIBQLGQbB1V1XPBZJsRwfK5r1U2Rzt"
-TEST_PUBLIC_PROFILE_ID="ACoAABQ11fIBQLGQbB1V1XPBZJsRwfK5r1U2Rzt"
-TEST_CONVERSATION_ID="6419123050114375168"
\ No newline at end of file
+LINKEDIN_USERNAME=test@mailnator.com
+LINKEDIN_PASSWORD=coolpasswordhere
+TEST_PROFILE_ID=ACoAABQ11fIBQLGQbB1V1XPBZJsRwfK5r1U2Rzt
+TEST_PUBLIC_PROFILE_ID=ACoAABQ11fIBQLGQbB1V1XPBZJsRwfK5r1U2Rzt
+TEST_CONVERSATION_ID=6419123050114375168
```
Comment (ssichynskyi) [inline .env.example:5]: remove double quotes that are both unnecessary and misleading
(cause issues at least on windows machine)

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -3,8 +3,10 @@
 1. Fork
 1. Write some code (with tests if possible)
 1. Write some docs
+1. Check code formatting.
+   In project root folder run `black --check`.
+   Proceed in case no issues found. Otherwise run `black .`
 1. Run the tests (with your `pipenv shell` activated)
```
Comment (ssichynskyi) [inline CONTRIBUTING.md:9]: add guideline to run formatting check / fix formatting

File: linkedin_api/linkedin.py [hunk 3]
```diff
@@ -144,9 +144,7 @@ def search(self, params, limit=-1, offset=0):
             # NOTE: we could also check for the `total` returned in the response.
             # This is in data["data"]["paging"]["total"]
             if (
-                (
-                    limit > -1 and len(results) >= limit
-                )  # if our results exceed set limit
+                (-1 < limit <= len(results))  # if our results exceed set limit
```
Comment (ssichynskyi) [inline linkedin_api/linkedin.py:147]: No real changes, just made the code a bit more beautiful

File: linkedin_api/linkedin.py [hunk 4]
```diff
@@ -326,8 +324,9 @@ def search_jobs(
         job_title=None,
         industries=None,
         location_name=None,
-        remote=True,
-        listed_at=86400,
+        remote=False,
```
Comment (ssichynskyi) [inline linkedin_api/linkedin.py:327]: The idea of "remote" field in query had changed. Now if remote=True, only jobs with remote working possibilities are listed. Therefore makes sense to make it "False" by default

File: linkedin_api/linkedin.py [hunk 5]
```diff
@@ -326,8 +324,9 @@ def search_jobs(
...
+        listed_at=24 * 60 * 60,
+        distance=None,
```
Comment (ssichynskyi) [inline linkedin_api/linkedin.py:329]: add distance param (how far from specified location job shall be searched)

File: linkedin_api/linkedin.py [hunk 6]
```diff
@@ -346,11 +345,18 @@ def search_jobs(
...
+        :param remote: Whether to search only for remote jobs. Defaults to False.
         :type remote: boolean, optional
-
+        :param listed_at: maximum number of seconds passed since job posting. 86400 will filter job postings posted in last 24 hours.
+        :type listed_at: int/str, optional. Default value is equal to 24 hours.
+        :param distance: maximum distance from location in miles
+        :type distance: int/str, optional. If not specified, None or 0, the default value of 25 miles applied.
+        :param limit: maximum number of results obtained from API queries. -1 means maximum which is defined by constants and is equal to 1000 now.
+        :type limit: int, optional, default -1
+        :param offset: indicates how many search results shall be skipped
```
Comment (ssichynskyi) [inline linkedin_api/linkedin.py:358]: add docu strings for new and old (but absent) parameters

File: linkedin_api/linkedin.py [hunk 7]
```diff
@@ -374,19 +380,28 @@ def search_jobs(
         if industries:
             filters.append(f'industry->{"|".join(industries)}')
         if location_name:
-            filters.append(f'locationFallback->{"|".join(location_name)}')
+            filters.append(f"locationFallback->{location_name}")
         if remote:
-            filters.append(f"commuteFeatures->f_WRA")
+            filters.append(f"workRemoteAllowed->{remote}")
```
Comment (ssichynskyi) [inline linkedin_api/linkedin.py:385]: fix of remote work allowed feature

File: linkedin_api/linkedin.py [hunk 8]
```diff
@@ -374,19 +380,28 @@ def search_jobs(
...
+        if distance:
+            filters.append(f"distance->{distance}")
```
Comment (ssichynskyi) [inline linkedin_api/linkedin.py:387]: add filtering by distance

File: linkedin_api/linkedin.py [hunk 9]
```diff
@@ -374,19 +380,28 @@ def search_jobs(
...
+        filters.append(f"timePostedRange->r{listed_at}")
+        # add optional kwargs to a filter
+        for name, value in kwargs.items():
+            if type(value) in (list, tuple):
+                filters.append(f'{name}->{"|".join(value)}')
+            else:
+                filters.append(f"{name}->{value}")
```
Comment (ssichynskyi) [inline linkedin_api/linkedin.py:394]: add handling of optional keyword arguments

File: linkedin_api/linkedin.py [hunk 10]
```diff
@@ -374,19 +380,28 @@ def search_jobs(
...
 
         results = []
         while True:
             # when we're close to the limit, only fetch what we need to
             if limit > -1 and limit - len(results) < count:
                 count = limit - len(results)
             default_params = {
-                "decorationId": "com.linkedin.voyager.deco.jserp.WebJobSearchHitWithSalary-14",
+                "decorationId": "com.linkedin.voyager.deco.jserp.WebJobSearchHitLite-14",
```
Comment (ssichynskyi) [inline linkedin_api/linkedin.py:402]: change api function. Might not be so important. I've just used one that is used by default in requests from linkedIn UI.

File: linkedin_api/linkedin.py [hunk 11]
```diff
@@ -374,19 +380,28 @@ def search_jobs(
...
                 "count": count,
-                "filters": f"List({filters})",
+                "filters": f"List({','.join(filters)})",
```
Comment (ssichynskyi) [inline linkedin_api/linkedin.py:404]: Critical change: fixes issue with no filters applied in job search.

File: linkedin_api/linkedin.py [hunk 12]
```diff
@@ -412,9 +427,7 @@ def search_jobs(
             # NOTE: we could also check for the `total` returned in the response.
             # This is in data["data"]["paging"]["total"]
             if (
-                (
-                    limit > -1 and len(results) >= limit
-                )  # if our results exceed set limit
+                (-1 < limit <= len(results))  # if our results exceed set limit
```
Comment (ssichynskyi) [inline linkedin_api/linkedin.py:430]: No real changes, just made the code a bit more beautiful

**Comment by ssichynskyi** [comment]
@tomquirk could you have a look please?

**Comment by tomquirk** [comment]
LGTM, can you fix conflicts @ssichynskyi ?

**Comment by ssichynskyi** [comment]
Ha-ha. This is an argument to not include pipfile.lock in repo. Although it's recommended from another reason. Tomorrow I'll re-do this from my Ubuntu.

**Comment by ssichynskyi** [comment]
@tomquirk, please merge the change. Conflicts are solved now.

**Comment by tomquirk** [comment]
>  This is an argument to not include pipfile.lock in repo.

https://developerexperience.io/practices/lockfile

Merging!

**Pull request merged**